### PR TITLE
Prevent Collection from adding 'rdf:nil rdf:rest rdf:nil.' triples

### PR DESCRIPTION
**Pull request by progval**
# Summary of changes

Creating 'Collection(graph, RDF.nil)' used to add a 'rdf:nil rdf:rest rdf:nil.' triple to the graph, which turned the empty list into an infinite list.

This was a side-effect of unconditionally appending 'seq' to the collection, which is not possible when the collection is 'nil' without ill side-effects.

This commit makes the append conditional; and also checks the collection is non-empty in case the user explicitly tries to add item to the empty list

# Checklist

- [x] Checked that there aren't other open pull requests for
  the same change.
- [ ] Checked that all tests and type checking passes.
- If the change has a potential impact on users of this project:
  - [x] Added or updated tests that fail without the change.
  - [ ] Updated relevant documentation to avoid inaccuracies.
  - [x] Considered adding additional documentation.
- [x] Considered granting [push permissions to the PR branch](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/working-with-forks/allowing-changes-to-a-pull-request-branch-created-from-a-fork),
  so maintainers can fix minor issues and keep your PR up to date.

### CODE REVIEW

File: rdflib/collection.py [hunk 1]
```diff
@@ -208,6 +212,8 @@ def __iter__(self) -> Iterator[Node]:
         return self.graph.items(self.uri)
 
     def _end(self) -> Node:
+        if self.uri == RDF.nil:
```
Comment (ashleysommer) [inline rdflib/collection.py:215]: Can we move this check from _end down to after line 252 in `__iadd__`, I think it makes more sense to catch it there.
That way we can still use `_end()` when `nil` is deliberately added to the end of a list.

**Review by ashleysommer — CHANGES_REQUESTED** [review]

File: rdflib/collection.py [hunk 2]
```diff
@@ -243,6 +250,9 @@ def append(self, item: Node) -> Collection:
         return self
 
     def __iadd__(self, other: Iterable[Node]):
+        if self.uri == RDF.nil:
+            raise ValueError("Cannot append to empty list")
+
         end = self._end()
```
Comment (ashleysommer) [inline rdflib/collection.py:256]: I was thinking like this:
```suggestion
        end = self._end()
        if end == RDF.nil:
            raise ValueError("Cannot append to empty list")
```

I think `_end()` will still return `nil` when `self.uri` is `nil`, this is what we want.

Note, I think you can append `nil` to the `rdf:rest` on the end of a populated list, that will cause it to be a finalized list, that cannot have more items added.

So if `_end()` returns `rdf:nil` then the list is either empty or finalized, either way it cannot have another item appended.

**Review by ashleysommer — CHANGES_REQUESTED** [review]

File: rdflib/collection.py [hunk 3]
```diff
@@ -231,6 +235,9 @@ def append(self, item: Node) -> Collection:
 
         """
 
+        if self.uri == RDF.nil:
+            raise ValueError("Cannot append to empty list")
+
         end = self._end()
```
Comment (ashleysommer) [inline rdflib/collection.py:241]: Can we make the same change here to `append()` too please?
```suggestion
        end = self._end()
        if end == RDF.nil:
            raise ValueError("Cannot append to empty list")

```

**Review by ashleysommer — CHANGES_REQUESTED** [review]

File: rdflib/collection.py (hunk 3 above)
Comment (progval) [reply rdflib/collection.py]: Sorry, I missed it

**Comment by ashleysommer** [comment]
There's an odd failing test now, weirdly its the same SPARQL test that is failing in #2796 that is completely unrelated to this PR.

**Comment by ashleysommer** [comment]
That sporadic test is no longer failing, this can merge.

**Review by ashleysommer — APPROVED** [review]

**Pull request merged**